geode-core/src/main/java/org/apache/geode/cache/query/internal/index/PartitionedIndex.java

```diff
@@ -175,13 +176,16 @@ public List getBucketIndexes() {
    */
   public Index getBucketIndex() {
     Index index = null;
-    synchronized (this.bucketIndexes) {
+    try {
       if (this.bucketIndexes.size() > 0) {
         List<Index> indexList = this.bucketIndexes.values().iterator().next();
         if (indexList != null && indexList.size() > 0) {
           index = indexList.get(0);
         }
       }
+    } catch (ConcurrentModificationException | NullPointerException e) {
```

Comment:
I assume the hope is that this catch will cover the case of an exception being thrown while iterating over the bucketIndexes synchronized map.  The thing that makes me nervous about this is that the documentation for synchronized maps says:
" It is imperative that the user manually synchronize on the returned map when iterating over any of its collection views
...
**Failure to follow this advice may result in non-deterministic behavior.** "

https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedMap%28java.util.Map%29

Reply:
I'm not sure if non-deterministic behavior is just limited to the occasional ConcurrentModificationException.

Reply:
I would be interested in knowing what other non-deterministic behavior can be exhibited.  Since this just wraps a basic HashMap, we can probably dig a bit and see what other exceptions are thrown...
For this specific usage, if the value no longer exists and we somehow retrieve it, that timing issue would exist anyways (imagine we retrieve it under a synchronize block but then soon after it's removed from the list).  The end result would be an index that no longer is present in the map/list combo but returned to the caller.
This method gets a bit of leeway as it's just picking an arbitrary index from this collection, other than other exceptions being thrown, retrieving something out of order isn't a big deal for this specific method.

In either case, we can improve performance even more by caching the arbitrary index.  So the non-determinism issue kind of goes away. for now... to be continued